CONFIG.md
# Configuration for `knitgateway`

This document describes the configuration options for the `knitgateway`
program.

There is a an example config file in the root of this repo named
[`knitgateway.example.yaml`](/knitgateway.example.yaml).
The example file shows all the properties that can be configured (though
most are commented out). It includes many comments to describe each
property.

The YAML file must contain only a single-document. If it contains any
unrecognized properties, the configuration will be rejected and the
gateway will not start.

There are several top-level keys in the file, each of which contains
properties for a different category of configuration.
1. **`listen`**: This section configures the listener for the gateway. This
   contains details about the server like what interface to bind to, what
   port to listen on, and whether TLS should be used.
2. **`limits`**: This section configures limits, to aid with operations.
3. **`backends`**: This section configures all of the backends to which the
   gateway can send requests when processing a Knit query. This section
   configures the available RPC services that can be used in a Knit query
   and how to route them to backends. It includes connectivity details but
   also details on how the gateway can access the schemata for the backend's
   exposed RPC services.
3. **`backend_tls`**: This section allows cross-cutting TLS configuration.
   Each backend can be separately configured in the `backends` section
   above. But if all or most backends use similar TLS configuration, it
   can be consolidated in this top-level section. When there is configuration
   in both this section and for a specific backend in `backends`, the values
   in `backends` _override_ the values here. So this section effectively
   defines the default TLS settings.
4. **`descriptors`**: This section controls the polling behavior of the
   gateway, for periodically reloading schemata. This allows the gateway
   to reconfigure itself at runtime as the schemas change. The schemas
   can also be cached, so a cached last-known-good schema can be used if
   the source of the schema is otherwise unavailable when the gateway
   starts up.
5. **`cors`**: This section controls how the gateway handles cross-origin
   requests and replies to CORS pre-flight requests.

Each of the above config stanzas is described thoroughly in the sections below.

## Listen Config

The top-level `listen` property is a _map_ with the following keys:

- **`bind_address`**: The address on which to listen. This defaults to 0.0.0.0,
  which means it will accept requests on _all_ network interfaces. You can
  specify a specific address to limit what interfaces are used. For example,
  setting this to 127.0.0.1 means that requests are only accepted on the
  loopback interface (i.e. from the local host).
- **`port`**: The port number on which to listen. There is no default: in order for
  the gateway to use a TCP listener, a port must be configured. If the port is
  configured as zero, an ephemeral port is used. In this case, the actual port
  in use will be logged when the gateway starts.
- **`unix_socket`**: The path to a Unix domain socket on which to listen. If the
  socket file already exists, the gateway will not use it and will not start. There
  is no default: in order for the gateway to listen on a Unix socket, this property
  must be configured.
- **`tls`**: If this section is present, the server will require clients to use
  TLS (transport-level security, sometimes called SSL) when connecting. This means
  that connections are secure: both parties can be authenticated via TLS certificates
  and all traffic is encrypted.

  This section is a map with the following sub-keys:
  - **`cert`**: This is the path to a PEM-encoded X509 certificate. This is required
    and configures the public key and certificate chain for the gateway's server
    certificate.
  - **`key`**: This is the path to a PEM-encoded X509 private key. This is required
    and configures the private key for the gateway's server certificate.
  - **`min_version`**: The minimum TLS version that the server will accept. This
    defaults to 1.0. Other allowed settings are 1.1, 1.2, or 1.3.
  - **`ciphers`**: The cipher suites to allow or disallow, for TLS 1.2 and below. (For
    TLS 1.3, the supported cipher suites are not configurable and are always the
    following: TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256.)

    This section is a map with possible keys of `allow` or `disallow`. Only one of these
    keys may be present, depending on whether the configuration is using an allow-list of
    cipher suites or a block-list. See [below](#tls-cipher-suites) for more details.
  - **`client_certs`**: If this section is present, the gateway will request TLS certificates
    from clients during the TLS handshake. This property is a map with the following allowed
    keys which further control the gateway's behavior.
    - **`require`**: If false or absent, client certs are verified if given, but are not
      required. If true, connections will be terminated if the client does not provide a
      valid cert.

      When a client cert is present and valid, the authenticated identity ("subject" field
      of the cert) will be added to HTTP headers for all requests to backends. It will be
      set in a header named `Knit-Client-Subject` and will be in RFC 2253 Distinguished Names
      syntax.
    - **`cacert`**: This is the path to a PEM-encoded X509 certificate pool file that contains
      certs for CAs (certificate authorities/issuers). These are used to verify client certs.

This property is required as the config _must_ define either a unix socket path or a
port (or both).

### TLS Cipher Suites

TLS cipher suites can be configured in one of two ways:
1. An allow-list is defined using the `allow` property. In this mode, only the suites
   listed in the config will be allowed.
2. A block-list is defined using the `disallow` property. In this mode, only the suites
   listed in the config will be blocked, and all others will be allowed.

The following table shows all supported cipher suites. The table also shows which suites
are allowed and disallowed by default, when no cipher suite configuration is provided,

| Cipher Suite                                  | TLS Versions | Disposition |
|-----------------------------------------------|--------------|-------------|
| TLS_RSA_WITH_AES_128_CBC_SHA                  |              | Allowed     |
| TLS_RSA_WITH_AES_256_CBC_SHA                  |              | Allowed     |
| TLS_RSA_WITH_AES_128_GCM_SHA256               | TLS 1.2 only | Allowed     |
| TLS_RSA_WITH_AES_256_GCM_SHA384               | TLS 1.2 only | Allowed     |
| TLS_RSA_WITH_AES_256_GCM_SHA384               | TLS 1.2 only | Allowed     |
| TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA          |              | Allowed     |
| TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA          |              | Allowed     |
| TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA            |              | Allowed     |
| TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA            |              | Allowed     |
| TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256       | TLS 1.2 only | Allowed     |
| TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384       | TLS 1.2 only | Allowed     |
| TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256         | TLS 1.2 only | Allowed     |
| TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384         | TLS 1.2 only | Allowed     |
| TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256   | TLS 1.2 only | Allowed     |
| TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256 | TLS 1.2 only | Allowed     |
| TLS_RSA_WITH_RC4_128_SHA                      |              | Disallowed  |
| TLS_RSA_WITH_3DES_EDE_CBC_SHA                 |              | Disallowed  |
| TLS_RSA_WITH_AES_128_CBC_SHA256               | TLS 1.2 only | Disallowed  |
| TLS_ECDHE_ECDSA_WITH_RC4_128_SHA              |              | Disallowed  |
| TLS_ECDHE_RSA_WITH_RC4_128_SHA                |              | Disallowed  |
| TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA           |              | Disallowed  |
| TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256       | TLS 1.2 only | Disallowed  |
| TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256         | TLS 1.2 only | Disallowed  |

## Limits Config

The top-level `limits` property is a map. At the moment, there is just one key
that is allowed.

- **`per_request_parallelism`**: Tte maximum parallelism per request. This is
  the maximum number of concurrent RPCs that can be made on behalf of a single
  Knit query.

  If not specified, there is no limit, and all RPCs needed to evaluate a query
  will all be made in parallel.

As other kinds of controls are implemented, configuration for them will be added
to this section.

## Backends Config

The top-level `backends` property is an _array_ of backend configuration maps. Each
map in the array has the following keys:

- **`route_to`**: The base URL for this backend. This key must be provided.
  The URL may use either "http" or "https" scheme. Note that no custom
  TLS properties (such as custom root CA certificate or client certificates)
  are currently supported for use with "https" URLs.
- **`unix_socket`**: If the backend is listening on a Unix domain socket and
  not a TCP socket, configure the path to the socket here. By default, TCP
  connections are used, but when this property is present and non-empty it
  is the path to a Unix socket that will be used instead.
- **`h2c`**: If the `routeTo` property uses a plaintext "http" scheme, but
  HTTP/2 should be used, set this property to true. It defaults to false.
- **`protocol`**: This configured the protocol that will be used to communicate
  with this backend. The default protocol is "connect". Other allowed options
  are "grpc" or "grpcweb".
- **`encoding`**: This configures the message encoding for sending requests to
  this backend. The default is "proto", which uses the Protobuf binary format.
  The other allowed option is "json".
- **`services`**: This key is required. The value is a list of fully-qualified
  service names that will be routed to this backend. If these services contain
  methods that can resolve relations, then those relations are automatically
  supported by the server.
- **`tls`**: This key configures TLS settings for the backend. This key may
  only appear when the `route_to` URL has a scheme of "https", indicating
  secure connections are used. This property is a map whose contents are used
  to configure a TLS client for successfully connecting to and verifying the
  backend. See [below](#tls-client-config) for more details about the format
  of this property.
- **`descriptors`**: This key is required. This configuration indicates how
  the gateway will find descriptors for the above named services. The value is
  a map with the following keys, of which _only one may be set_:
    - **`descriptor_set_file`**: The value for this key is the path to a file
      that is an encoded [file descriptor set](https://github.com/protocolbuffers/protobuf/blob/v22.0/src/google/protobuf/descriptor.proto#L54-L58).
      Both `buf` and `protoc` can produce such files. (See more [below](#descriptor-set-examples).)
    - **`buf_module`**: The value for this key is the name of a module that has
      been pushed to a Buf Schema Registry (BSR). The module name must be in the
      format "&lt;remote>/&lt;owner>/&lt;repo>". The first part defines the host name for
      the BSR, for example `buf.build`. When using this option, you must provide
      an environment variable named [`BUF_TOKEN`](https://docs.buf.build/bsr/authentication#buf_token)
      that the gateway will use to authenticate with the BSR in order to download
      the module's descriptors.
    - **`grpc_reflection`**: The value for this key is a boolean. If true, then
      the [gRPC Server Reflection](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md)
      protocol will be used to download the descriptors from the
      backend server itself.

### TLS Client Config

Each backend configured in the `backends` stanza can configure TLS client settings
via a `tls` property. Default TLS settings for all secure backends can be defined
in the top-level `backend_tls` stanza. No TLS settings are required; reasonable
defaults will be used for all settings.

The property is a map with the following keys:

- **`min_version`**: This minimum version of TLS to accept. Defaults to 1.2. Other
  allowed values are 1.1, 1.2, and 1.3. Note that the default here is different than
  for TLS _server_ settings, in the `tls` property of the [`listen`](#listen-config)
  stanza. The listener default is more lenient, to support external clients that may
  be using older browser or mobile OS software.
- **`ciphers`**: The cipher suites to allow or disallow, for TLS 1.2 and below. (For
  TLS 1.3, the supported cipher suites are not configurable and are always the
  following: TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256.)

  This section is a map with possible keys of `allow` or `disallow`. Only one of these
  keys may be present, depending on whether the configuration is using an allow-list of
  cipher suites or a block-list. See [below](#tls-cipher-suites) for more details.
- **`cacert`**: This is the path to a PEM-encoded X509 certificate pool file that
  contains certs for CAs (certificate authorities/issuers). These are used to verify
  server certs of the backends. If not present or blank, system defaults for the set
  of trusted root CAs will be used.
- **`skip_verify`**: This flag **disables** verification of server certs. Its use is
  strongly discouraged. It is present primarily to aid with testing.
- **`cert`**: This is the path to a PEM-encoded X509 certificate. This configures
  the public key and certificate chain for the client certificate to use. This should
  only be present if the server expects clients to provide a TLS certificate. If this
  property is present, `key` must also be present.
- **`key`**: This is the path to a PEM-encoded X509 private key. This configures
  the private key for a client certificate to use. This should only be present if the
  server expects clients to provide a TLS certificate. If this property is present,
  `cert` must also be present.

### Descriptor Set Examples

If using the `descriptor_set_file` option for the `descriptors` key, you can
use `buf` or `protoc` to generate a file in the correct format.

The following example uses `buf` to build proto sources in the current
directory. The `-o` flag indicates the path to the output file that will
contain the descriptors:

```shell
buf build . -o ../my-services.protoset
```

Here's another example using `buf`, this time to build the
[buf.build/bufbuild/knit-demo](https://buf.build/bufbuild/knit-demo) module.
This module contains service definitions that describe the
_Star Wars API_ ([swapi.dev](https://swapi.dev/)).

```shell
buf build buf.build/bufbuild/knit-demo -o swapi.protoset
```

Finally, here's an example that uses `protoc`. This compiles a
hypothetical file at path `foo/bar/services.proto`, where some of
its imports are defined in a `../../others/proto` directory.

Here too, the `-o` flag indicates the path to the output file that will
contain the descriptors. Most importantly, you must also include the
`--include_imports` flag, or else the resulting file may be incomplete
and unusable by the Knit gateway.

```shell
protoc -I ../../others/proto foo/bar/services.proto \
    --include_imports -o ../my-services.protoset
```

Note that using `protoc` may involve specifying multiple input files
and multiple `-I` include path options. Refer to your existing scripts
that invoke `protoc` for code generation.

## Backend TLS Config

The top-level `backend_tls` property is a map that defines default TLS
settings. These settings are used with any backend with a secure URL
(i.e. scheme is "https"). These settings may be overridden on a per-backend
basis via the `tls` property for that backend.

The allowed keys in this map are the same as for the `tls` property for a
backend, [as described above](#tls-client-config). There is one exception:
the `backend_tls` map may _not_ include a `server_name` key. An override
server name can only be configured on a per-backend basis and cannot be
set in the defaults.

## Descriptors Config

The top-level `descriptors` property is a map that defines default settings
for polling and caching of descriptors, which define the schemas for the
supported RPC services. It allows the following keys:

- **`startup_max_wait_seconds`**: This is the maximum time to wait at startup
  for schemas to be resolved and all descriptors to be downloaded. If it takes
  longer than this to resolve schemas, the process will exit with an error
  instead of continuing to wait.

  If unset or zero, then a default value of 15 seconds is used.
- **`polling_period_seconds`**: The time to wait in between attempts to
  re-download descriptors. The gateway continually re-downloads schemas in case
  they change over time.

  If unset or set to zero, a default value of 15 minutes is used (which is 900
  seconds).
- **`polling_jitter`**: A value between zero and one for the amount of random
  jitter to use when scheduling a polling attempt. This is used to prevent
  multiple processes from inadvertently self-synchonizing and turning into a
  thundering herd. A typical value for this purpose is 0.1 to 0.3.

  A value of zero means no jitter. A value of 1 means 100% jitter, which means
  the polling period can be perturbed up to 100% (so it could be as low as zero
  as high as double the configured period). The default is 0.25.
- **`polling_debounce_seconds`**: The number of seconds to wait after a schema
  update to "debounce" updates from multiple sources.

  A smaller value means the gateway's internals are re-created more frequently
  when updates are frequent. A higher value can be more efficient as it re-creates
  the internals only once for a sequence of rapid updates, but it may slow down
  the reaction time from receiving a new schema and serving the corresponding new
  configuration. Note that the server could benefit from debouncing even when the
  polling period is high because there could be multiple sources of descriptors,
  so multiple updates could be arriving, all from a single scheduled re-polling of
  descriptors.

  The default is zero, which is appropriate for development. But production
  deployments with multiple schema sources should consider setting it to a value
  to prevent too much CPU time being used by re-creating configuration. Between
  5 ad 30 seconds is a reasonable range of values.
- **`cache`**: This section configures a cache for resolved schemas. For Buf BSR
  modules and gRPC reflection as descriptor sources, there is a possibility that
  a network partition could prevent the gateway from downloading a schema at

internal/app/knitgateway/config.go
// Copyright 2023 Buf Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package knitgateway

import (
	"context"
	"crypto/tls"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"time"

	connect "github.com/bufbuild/connect-go"
	"github.com/bufbuild/knit-go"
	"github.com/bufbuild/prototransform"
	"github.com/rs/cors"
	"golang.org/x/net/http2"
	"gopkg.in/yaml.v3"
)

const (
	httpScheme  = "http"
	httpsScheme = "https"

	connectProtocol = "connect"
	grpcProtocol    = "grpc"
	grpcWebProtocol = "grpcweb"

	protoEncoding = "proto"
	jsonEncoding  = "json"

	tlsVersion10 = "1.0"
	tlsVersion11 = "1.1"
	tlsVersion12 = "1.2"
	tlsVersion13 = "1.3"

	defaultPollingPeriod  = 15 * time.Minute
	defaultStartupMaxWait = 15 * time.Second
)

// GatewayConfig is the configuration for a Knit gateway.
type GatewayConfig struct {
	ListenAddress            string
	UnixSocket               string
	TLSConfig                *tls.Config
	CORSConfig               cors.Options
	Services                 map[string]ServiceConfig
	MaxParallelismPerRequest int
	StartupMaxWait           time.Duration
	PollingPeriod            time.Duration
	PollingJitter            float64
	PollingDebounce          time.Duration
	Cache                    DescriptorCacheConfig
}

// ServiceConfig is the configuration for a single RPC service.
type ServiceConfig struct {
	BaseURL     *url.URL
	HTTPClient  connect.HTTPClient
	ConnectOpts []connect.ClientOption
	Descriptors prototransform.SchemaPoller
	Cacheable   bool
}

// LoadConfig reads the config file at the given path and returns the resulting
// GatewayConfig on success.
func LoadConfig(path string) (*GatewayConfig, error) { //nolint:gocyclo
	configInput, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer func() {
		_ = configInput.Close()
	}()
	dec := yaml.NewDecoder(configInput)
	dec.KnownFields(true)
	var extConf externalGatewayConfig
	if err := dec.Decode(&extConf); err != nil {
		return nil, err
	}
	var dummy any
	if err := dec.Decode(&dummy); !errors.Is(err, io.EOF) {
		return nil, errors.New("config file contains multiple documents, should only contain one")
	}

	if extConf.Listen.BindAddress == "" {
		extConf.Listen.BindAddress = "0.0.0.0"
	}
	if extConf.Listen.Port == nil && extConf.Listen.UnixSocket == "" {
		return nil, errors.New("listen config must include at least one of 'port' or 'unix_socket'")
	}

	if len(extConf.Backends) == 0 {
		return nil, fmt.Errorf("backends config is empty")
	}

	pollers := map[descriptorSource]prototransform.SchemaPoller{}
	numSourcesCacheable := 0
	backendConfigIndexByService := map[string]int{}
	services := map[string]ServiceConfig{}

	serverTLS, err := configureServerTLS(extConf.Listen.TLS)
	if err != nil {
		return nil, fmt.Errorf("listen.tls: invalid configuration: %w", err)
	}
	// we use this to validate the default_tls config (and can discard the result)
	_, err = configureClientTLS(extConf.DefaultBackendTLS)
	if err != nil {
		return nil, fmt.Errorf("default_tls: invalid configuration: %w", err)
	}
	if extConf.DefaultBackendTLS != nil && extConf.DefaultBackendTLS.ServerName != "" {
		return nil, fmt.Errorf("default_tls: invalid configuration: server_name property cannot be used in default settings, only in per-backend config")
	}

	for i, backendConf := range extConf.Backends {
		if backendConf.RouteTo == "" {
			return nil, fmt.Errorf("backend config #%d: missing 'route_to' property", i+1)
		}
		if len(backendConf.Services) == 0 {
			return nil, fmt.Errorf("backend config #%d: 'services' property should not be empty", i+1)
		}
		routeURL, err := url.Parse(backendConf.RouteTo)
		if err != nil {
			return nil, fmt.Errorf("backend config #%d: %q is not a valid URL: %w", i+1, backendConf.RouteTo, err)
		}
		if routeURL.Scheme != httpScheme && routeURL.Scheme != httpsScheme {
			return nil, fmt.Errorf("backend config #%d: %q is not a valid URL: schema should be '%s' or '%s'",
				i+1, backendConf.RouteTo, httpScheme, httpsScheme)
		}
		var options []connect.ClientOption
		switch backendConf.Encoding {
		case "", protoEncoding:
			// nothing to do; this is the default
		case jsonEncoding:
			options = append(options, knit.WithProtoJSON())
		default:
			return nil, fmt.Errorf("backend config #%d: %q is not a valid encoding; should be '%s' or '%s'",
				i+1, backendConf.Encoding, protoEncoding, jsonEncoding)
		}
		h2c := backendConf.H2C != nil && *backendConf.H2C
		if h2c && routeURL.Scheme == httpsScheme {
			return nil, fmt.Errorf("backend config #%d: 'h2c' is enabled, but URL scheme is %q; H2C is for plaintext connections", i+1, routeURL.Scheme)
		}
		switch backendConf.Protocol {
		case "", connectProtocol:
			// nothing to do; this is the default
		case grpcProtocol:
			options = append(options, connect.WithGRPC())
			// for grpc over plaintext, we must use H2C
			if backendConf.H2C == nil {
				h2c = routeURL.Scheme == httpScheme
			} else if !h2c {
				// H2C is explicitly disabled, which is not valid configuration
				return nil, fmt.Errorf("backend config #%d: cannot use %s protocol with %s schema w/out H2C",
					i+1, grpcProtocol, httpScheme)
			}
		case grpcWebProtocol:
			options = append(options, connect.WithGRPCWeb())
		default:
			return nil, fmt.Errorf("backend config #%d: %q is not a valid protocol; should be '%s', '%s', or '%s'",
				i+1, backendConf.Protocol, connectProtocol, grpcProtocol, grpcWebProtocol)
		}
		var clientTLS *tls.Config
		if routeURL.Scheme != httpsScheme {
			if backendConf.TLS != nil {
				return nil, fmt.Errorf("backend config #%d: tls stanza present but URL scheme is not https", i+1)
			}
		} else {
			clientTLS, err = configureClientTLS(mergeClientTLSConfig(backendConf.TLS, extConf.DefaultBackendTLS))
			if err != nil {
				return nil, fmt.Errorf("backend config #%d: tls: invalid configuration: %w", i+1, err)
			}
			if clientTLS != nil && backendConf.TLS != nil && backendConf.TLS.ServerName != "" {
				clientTLS.ServerName = backendConf.TLS.ServerName
			}
		}

		httpClient, err := makeHTTPClient(clientTLS, backendConf.UnixSocket, h2c)
		if err != nil {
			return nil, fmt.Errorf("backend config #%d: failed to create HTTP client: %w", i+1, err)
		}
		svcConf := ServiceConfig{
			BaseURL:     routeURL,
			HTTPClient:  httpClient,
			ConnectOpts: options,
		}
		descSrc, err := newDescriptorSource(backendConf.Descriptors, backendConf.RouteTo, h2c)
		if err != nil {
			return nil, fmt.Errorf("backend config #%d: %w", i+1, err)
		}
		if descSrc.isCacheable() {
			numSourcesCacheable++
			svcConf.Cacheable = true
		}
		poller := pollers[descSrc]
		if poller == nil {
			poller, err = descSrc.newPoller(httpClient, options)
			if err != nil {
				return nil, fmt.Errorf("backend config #%d: %w", i+1, err)
			}
			pollers[descSrc] = poller
		}
		svcConf.Descriptors = poller

		for _, svc := range backendConf.Services {
			if otherIndex, ok := backendConfigIndexByService[svc]; ok {
				return nil, fmt.Errorf("backend config #%d: service %q is already configured under backend config #%d", i+1, svc, otherIndex+1)
			}
			backendConfigIndexByService[svc] = i
			services[svc] = svcConf
		}
	}

	var listenAddress string
	if extConf.Listen.Port != nil {
		listenAddress = fmt.Sprintf("%s:%d", extConf.Listen.BindAddress, *extConf.Listen.Port)
	}

	pollingPeriod := defaultPollingPeriod
	if extConf.Descriptors.PollingPeriodSeconds < 0 {
		return nil, fmt.Errorf("descriptors.polling_period_seconds may not be negative: %d", extConf.Descriptors.PollingPeriodSeconds)
	}
	if extConf.Descriptors.PollingPeriodSeconds > 0 {
		pollingPeriod = time.Second * time.Duration(extConf.Descriptors.PollingPeriodSeconds)
	}

	jitter := 0.25
	if extConf.Descriptors.PollingJitter != nil {
		jitter = *extConf.Descriptors.PollingJitter
		if jitter < 0 {
			return nil, fmt.Errorf("descriptors.polling_jitter may not be negative: %f", jitter)
		}
		if jitter > 1 {
			return nil, fmt.Errorf("descriptors.polling_jitter may not be greater than 1.0: %f", jitter)
		}
	}

	if extConf.Descriptors.PollingDebounceSeconds < 0 {
		return nil, fmt.Errorf("descriptors.polling_debounce_seconds may not be negative: %d", extConf.Descriptors.PollingDebounceSeconds)
	}

	startupMaxWait := defaultStartupMaxWait
	if extConf.Descriptors.StartupMaxWaitSeconds < 0 {
		return nil, fmt.Errorf("descriptors.startup_max_wait_seconds may not be negative: %d", extConf.Descriptors.StartupMaxWaitSeconds)
	}
	if extConf.Descriptors.StartupMaxWaitSeconds > 0 {
		startupMaxWait = time.Second * time.Duration(extConf.Descriptors.StartupMaxWaitSeconds)
	}

	cacheConfig, err := newCacheConfig(extConf.Descriptors.Cache, numSourcesCacheable)
	if err != nil {
		return nil, err
	}

	corsConfig := cors.Options{
		AllowedOrigins:      extConf.CORS.AllowedOrigins,
		AllowedHeaders:      extConf.CORS.AllowedHeaders,
		AllowCredentials:    extConf.CORS.AllowCredentials,
		AllowPrivateNetwork: extConf.CORS.AllowPrivateNetworks,
		MaxAge:              extConf.CORS.MaxAgeSeconds,
	}
	if len(corsConfig.AllowedOrigins) == 0 {
		// If allowed origina is empty, the cors library defaults to allowing all '*'.
		// But we want users to opt into that behavior. So if the list is empty, we
		// provide a function that will be called (instead of using the empty list
		// of origins).
		corsConfig.AllowOriginFunc = func(string) bool { return false }
	}

	return &GatewayConfig{
		ListenAddress:            listenAddress,
		UnixSocket:               extConf.Listen.UnixSocket,
		TLSConfig:                serverTLS,
		CORSConfig:               corsConfig,
		Services:                 services,
		MaxParallelismPerRequest: extConf.Limits.PerRequestParallelism,
		StartupMaxWait:           startupMaxWait,
		PollingPeriod:            pollingPeriod,
		PollingJitter:            jitter,
		PollingDebounce:          time.Second * time.Duration(extConf.Descriptors.PollingDebounceSeconds),
		Cache:                    cacheConfig,
	}, nil
}

type externalGatewayConfig struct {
	Listen            externalListenConfig            `yaml:"listen"`
	Backends          []externalBackendConfig         `yaml:"backends"`
	Limits            externalLimitsConfig            `yaml:"limits"`
	DefaultBackendTLS *externalClientTLSConfig        `yaml:"backend_tls"`
	CORS              externalCORSConfig              `yaml:"cors"`
	Descriptors       externalDescriptorPollingConfig `yaml:"descriptors"`
}

type externalListenConfig struct {
	BindAddress string `yaml:"bind_address"`
	Port        *int   `yaml:"port"`
	// If present, gateway will listen for requests on unix domain socket with given path.
	// This is in addition to TCP listener (unless port is missing, in which case only the
	// unix socket listener is used).
	UnixSocket string `yaml:"unix_socket"`
	// If 'tls' section is present, clients must use TLS when connecting.
	TLS *externalServerTLSConfig `yaml:"tls"`
}

type externalServerTLSConfig struct {
	// Cert and Key are both required.
	Cert string `yaml:"cert"`
	Key  string `yaml:"key"`
	// If absent, defaults to "1.0". Allowed values are "1.0", "1.1", "1.2", and "1.3".
	MinVersion string `yaml:"min_version"`
	// If absent, a default set of secure cipher suites is used.
	Ciphers *externalCiphersConfig `yaml:"ciphers"`
	// If 'client_certs' key is absent, client TLS certs will be ignored and not validated.
	ClientCerts *externalClientCertsConfig `yaml:"client_certs"`
}

type externalCiphersConfig struct {
	// Exactly one of the following fields may be used, not both.

	// Enumerates the allowed cipher suite names.
	Allow []string `yaml:"allow"`
	// Enumerates the disallowed cipher suite names. The default cipher suites will be used
	// but items named here will be excluded.
	Disallow []string `yaml:"disallow"`
}

type externalClientCertsConfig struct {
	Require bool   `yaml:"require"`
	CACert  string `yaml:"cacert"`
}

type externalBackendConfig struct {
	RouteTo string `yaml:"route_to"`

	// If present, requests will be sent via unix socket instead of TCP connection.
	UnixSocket string `yaml:"unix_socket"`
	// Should not be present unless 'route_to' indicates "https" scheme. Only
	// required if client certs are to be used, if a custom CA is needed for
	// verifying backend cert, a custom server-name is needed (for SNI), or if
	// top-level 'default_tls' config needs to be overridden.
	TLS *externalClientTLSConfig `yaml:"tls"`

	H2C         *bool                    `yaml:"h2c"`
	Protocol    string                   `yaml:"protocol"`
	Encoding    string                   `yaml:"encoding"`
	Descriptors externalDescriptorConfig `yaml:"descriptors"`
	Services    []string                 `yaml:"services"`
}

type externalClientTLSConfig struct {
	// If absent, defaults to "1.2". Allowed values are "1.0", "1.1", "1.2", and "1.3".
	MinVersion *string `yaml:"min_version"`
	// If absent, a default set of secure cipher suites is used.
	Ciphers *externalCiphersConfig `yaml:"ciphers"`

	CACert *string `yaml:"cacert"`
	// If true, the backend's certificate is not verified. For testing only.
	// It is STRONGLY discouraged to use this flag in real production environments.
	// If present, any value for 'cacert' is ignored.
	SkipVerify *bool `yaml:"skip_verify"`

	// If one of the following is present, then both must be present. When present,
	// they will be used as the client certificated presented to the backend server
	// during TLS handshake.
	Cert *string `yaml:"cert"`
	Key  *string `yaml:"key"`

	// Override server name, for SNI.
	ServerName string `yaml:"server_name"`

	// !!NOTE: If any other fields are added, also update mergeClientTLSConfig in tls.go!!!
}

type externalDescriptorConfig struct {
	DescriptorSetFile string `yaml:"descriptor_set_file"`
	BufModule         string `yaml:"buf_module"`
	// TODO: Headers/authn for reflection requests.
	GRPCReflection bool `yaml:"grpc_reflection"`
}

type externalDescriptorPollingConfig struct {
	StartupMaxWaitSeconds  int                           `yaml:"startup_max_wait_seconds"`
	PollingPeriodSeconds   int                           `yaml:"polling_period_seconds"`
	PollingJitter          *float64                      `yaml:"polling_jitter"`
	PollingDebounceSeconds int                           `yaml:"polling_debounce_seconds"`
	Cache                  externalDescriptorCacheConfig `yaml:"cache"`
}

type externalLimitsConfig struct {
	PerRequestParallelism int `yaml:"per_request_parallelism"`
	// TODO: add other limits like message read size limits, resolver batch size limits,
	//       max total parallelism, max concurrent requests, rate limits, etc.
}

type externalCORSConfig struct {
	AllowedOrigins       []string `yaml:"allowed_origins"`
	AllowedHeaders       []string `yaml:"allowed_headers"`
	AllowCredentials     bool     `yaml:"allow_credentials"`
	AllowPrivateNetworks bool     `yaml:"allow_private_networks"`
	MaxAgeSeconds        int      `yaml:"max_age_seconds"`
}

func makeHTTPClient(tlsConfig *tls.Config, unixSocket string, h2c bool) (connect.HTTPClient, error) {
	if tlsConfig == nil && unixSocket == "" {
		// can use default clients
		if h2c {
			return defaultH2CClient, nil
		}
		return http.DefaultClient, nil
	}

	dial := defaultDialer.DialContext
	if unixSocket != "" {
		if err := checkUnixSocket(unixSocket); err != nil {
			return nil, errorHasFilename(err, unixSocket)
		}
		dial = func(ctx context.Context, _, _ string) (net.Conn, error) {
			return defaultDialer.DialContext(ctx, "unix", unixSocket)
		}
	}

	if h2c {
		// This is the same as used above in defaultH2CClient
		transport := &http2.Transport{
			AllowHTTP: true,
			DialTLSContext: func(ctx context.Context, network, addr string, _ *tls.Config) (net.Conn, error) {
				return dial(ctx, network, addr)
			},
		}
		if tlsConfig != nil {
			transport.TLSClientConfig = tlsConfig
		}
		return &http.Client{Transport: transport}, nil
	}

	// This is the same as http.DefaultTransport.
	transport := &http.Transport{
		Proxy:                 http.ProxyFromEnvironment,
		DialContext:           dial,
		ForceAttemptHTTP2:     true,
		MaxIdleConns:          100,
		IdleConnTimeout:       90 * time.Second,
		TLSHandshakeTimeout:   10 * time.Second,
		ExpectContinueTimeout: 1 * time.Second,
	}
	if tlsConfig != nil {
		transport.TLSClientConfig = tlsConfig
	}
	return &http.Client{Transport: transport}, nil
}

func checkUnixSocket(socketPath string) error {
	conn, err := defaultDialer.Dial("unix", socketPath)
	if err != nil {
		return err
	}
	return conn.Close()
}
